### ISSUE TITLE
Support for external database

### ISSUE THREAD

**Issue by djjudas21** [navidrome/navidrome#1975]
### Is your feature request related to a problem? Please describe.

I would like the ability to use an external database instead of SQLite, e.g. PostgreSQL or MariaDB.

In my case, I am running Navidrome in Kubernetes and the only persistent storage available is an NFS-backed storage class. Only Navidrome accesses the SQLite databases, but quite frequently locks aren't released properly and Navidrome errors because the database is locked. Usually I have to restart the Navidrome pod when this happens.

I do have fast and reliable PostgreSQL and MariaDB databases available in my environment, so I would prefer to use this for Navidrome's database instead of SQLite.

### Describe the solution you'd like

I don't know how database support is implemented in Navidrome, but I guess it would be possible to use a database abstraction layer which would give the user a huge choice of database backends. The default could remain as SQLite

### Describe alternative solutions that would also satisfy this problem

My NAS and the Kubernetes StorageClass are using NFSv3. I plan to migrate to NFSv4 which has better support for locking, so maybe this will mitigate my issue.

**Comment by deluan** [comment]
> but I guess it would be possible to use a database abstraction layer which would give the user a huge choice of database backends. The default could remain as SQLite

That was the original idea, this layer is in the `persistence` package. But as the project grew, I had to make optimizations specific to SQLite, even using some SQLite3-only features to implement parts of the code (ex: the scrobble queue, and soon the full text search). 

So this is not feasible without a big refactor, unfortunately. I'll keep this open for now, but if I feel this will never happen, I'll let you know.

Re: SQLite over NFS (or any networked drive for that matter), this is known to NOT work properly, and the SQLite documentation explicitly [says it should be avoid](https://sqlite.org/lockingv3.html) (search for _Your best defense is to not use SQLite for files on a network filesystem_)

**Comment by djjudas21** [comment]
Thanks for the info @deluan. Yes I know SQLite over NFS is an anti-pattern, but they do also say with a single client, it should be "safe". So I'm not sure why I'm running into locking problems, but I will see if NFSv4 improves it.

**Comment by deluan** [comment]
Hey @djjudas21, did you get to test it with NFSv4?

I've been refactoring the persistence layer of Navidrome, with the goal of (someday) be able to allow different DB engines. But for now, can we close this?

**Comment by djjudas21** [comment]
I didn't manage to do a conclusive test with NFSv4. I enabled NFSv4 support on my NAS and CSI provisioner, but this only affects creation of new NFS volumes, it doesn't affect the way existing NFSv3 volumes are mounted.

I have migrated my Navidrome databases to a different storageclass (CStor hyperconverged NVMe, presented as iSCSI). This works much better, so my immediate problem has been fixed.

I guess it's fine to close this issue, and if anyone else searches for this, hopefully they'll find this thread with the advice to avoid NFS if possible :grinning:
